src/Xamarin.Android.Build.Tasks/Utilities/AssemblyPipeline.cs

```diff
@@ -0,0 +1,76 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using Java.Interop.Tools.Cecil;
+using Java.Interop.Tools.JavaCallableWrappers;
+using Microsoft.Build.Framework;
+using Mono.Cecil;
+
+namespace Xamarin.Android.Tasks;
+
+public class AssemblyPipeline : IDisposable
+{
+	bool disposed_value;
+
+	public List<IAssemblyModifierPipelineStep> Steps { get; } = [];
+	public DirectoryAssemblyResolver Resolver { get; }
+
+	public AssemblyPipeline (DirectoryAssemblyResolver resolver)
+	{
+		Resolver = resolver;
+	}
+
+	public bool Run (AssemblyDefinition assembly, StepContext context)
+	{
+		var changed = false;
+
+		foreach (var step in Steps)
+			changed |= step.ProcessAssembly (assembly, context);
+
+		return changed;
+	}
+
+	protected virtual void Dispose (bool disposing)
+	{
+		if (!disposed_value) {
+			if (disposing) {
+				Resolver?.Dispose ();
+			}
+
+			disposed_value = true;
+		}
+	}
+
+	public void Dispose ()
+	{
+		// Do not change this code. Put cleanup code in 'Dispose (bool disposing)' method
+		Dispose (disposing: true);
+		GC.SuppressFinalize (this);
+	}
+}
+
+public interface IAssemblyModifierPipelineStep
+{
+	bool ProcessAssembly (AssemblyDefinition assembly, StepContext context);
+}
```

Comment:
Could this go in its own file, so then `src\Microsoft.Android.Sdk.ILLink\Microsoft.Android.Sdk.ILLink.csproj` could use it?